cmd/query/app/server_test.go

```diff
@@ -103,3 +103,23 @@ func TestServerGracefulExit(t *testing.T) {
 			"Error log found on server exit: %v", logEntry)
 	}
 }
+
+func TestServerHandlesPortZero(t *testing.T) {
+	flagsSvc := flags.NewService(ports.QueryAdminHTTP)
+	zapCore, logs := observer.New(zap.InfoLevel)
+	flagsSvc.Logger = zap.New(zapCore)
+
+	querySvc := &querysvc.QueryService{}
+	tracer := opentracing.NoopTracer{}
+	server := NewServer(flagsSvc, querySvc, &QueryOptions{Port: 0}, tracer)
+	assert.NoError(t, server.Start())
+	server.Close()
+
+	message := logs.FilterMessage("Query server started")
+	assert.Equal(t, 1, message.Len(), "Expected query started log message.")
+
+	onlyEntry := message.All()[0]
+	port := onlyEntry.ContextMap()["port"].(int)
+	assert.True(t, port > 0,
+		"Expected a non-zero port in the logs, got instead: %d", port)
```

Comment:
strictly speaking, it is possible that NewServer() function replaces 0 port with the default value, in which case this test will not catch it.

Reply:
one way to guarantee that it does is by starting two servers and validating that they both use different ports.

Reply:
can we fix this?

Reply:
> strictly speaking, it is possible that NewServer() function replaces 0 port with the default value, in which case this test will not catch it.

> one way to guarantee that it does is by starting two servers and validating that they both use different ports.


Sure, I will add that test. However, what this PR targets is printing the real port in the log file instead of 0 and not the fact of using port 0 itself.
Using 0 works even before this PR :-)


btw, I also have a gap for testing the admin port log output. In the current state, I do not have a way to set the logger on the `AdminServer`. What I can do is to expose `setLogger()` and use it, then call `Serve()` and check the log in that case. Would that be ok?

Reply:
Logger can be passed via InitFromViper. Look for Viperize helper function somewhere in the repo.

Reply:
I added tests for launching two instances in parallel. I used a cyclicbarrier to make sure that both servers are synchronized after the Start(), because without it, each goroutine can run on its own and the test behavior will be random. Let me know if you are ok with the thirdparty, or if you have other suggestions. :-)